src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
         return ret;
     }
 
+    private class RegistryCodec implements Codec<V> {
+        @Override
+        public <T> DataResult<Pair<V, T>> decode(DynamicOps<T> ops, T input) {
+            if(ops.compressMaps())
+            {
+                return ops.getNumberValue(input).flatMap(n ->
+                {
+                    int id = n.intValue();
+                    if(ids.get(id) == null)
+                    {
+                        return DataResult.error("Unknown registry id " + n);
+                    }
+                    V val = ForgeRegistry.this.getValue(id);
+                    return DataResult.success(val);
+                }).map(v -> Pair.of(v, input));
```

Comment:
The `Pair` returned by `#decode` represents the data that was successfully decoded to the object and the data that failed to decode. This says that the output was both fully encoded and decoded like schrodinger's cat.

Reply:
You're right, vanilla registry codec return ops.empty() here...